File: avalon-modules/avalon-authorization/src/main/java/io/github/lishangbu/avalon/authorization/repository/MenuRepository.java
package io.github.lishangbu.avalon.authorization.repository;

import io.github.lishangbu.avalon.authorization.entity.Menu;
import java.util.List;
import org.springframework.data.jpa.repository.JpaRepository;
import org.springframework.data.jpa.repository.JpaSpecificationExecutor;
import org.springframework.data.jpa.repository.Query;
import org.springframework.data.repository.query.Param;
import org.springframework.stereotype.Repository;

/// 菜单表数据存储
///
/// 提供基础的 CRUD 操作
///
/// @author lishangbu
/// @since 2025/08/20
@Repository
public interface MenuRepository extends JpaRepository<Menu, Long>, JpaSpecificationExecutor<Menu> {
    /// 根据角色代码列表查询对应的菜单集合
    ///
    /// @param roleCodes 角色代码列表
    /// @return 匹配的菜单列表
    @Query(
            """
                    select distinct m from Role r
                    join r.menus m
                    where r.code in :roleCodes
                    order by m.sortingOrder desc
                    """)
    List<Menu> findAllByRoleCodes(@Param("roleCodes") List<String> roleCodes);
}

File: avalon-modules/avalon-authorization/src/main/kotlin/io/github/lishangbu/avalon/authorization/repository/MenuRepository.kt
package io.github.lishangbu.avalon.authorization.repository

import io.github.lishangbu.avalon.authorization.entity.*
import io.github.lishangbu.avalon.authorization.entity.dto.MenuTreeView
import io.github.lishangbu.avalon.authorization.entity.dto.MenuView
import org.babyfish.jimmer.Specification
import org.babyfish.jimmer.spring.repository.KRepository
import org.babyfish.jimmer.spring.repository.orderBy
import org.babyfish.jimmer.sql.kt.ast.expression.eq
import org.babyfish.jimmer.sql.kt.ast.expression.isNull
import org.springframework.data.domain.Sort

/**
 * 菜单仓储接口
 *
 * 定义菜单数据的查询与持久化操作
 *
 * @author lishangbu
 * @since 2025/08/20
 */
interface MenuRepository : KRepository<Menu, Long> {
    /** 按条件查询菜单列表 */
    fun findAll(specification: Specification<Menu>?): List<Menu> =
        sql
            .createQuery(Menu::class) {
                specification?.let(::where)
                select(table.fetch(AuthorizationFetchers.MENU))
            }.execute()

    /** 按条件查询菜单视图列表 */
    fun listViews(specification: Specification<Menu>?): List<MenuView> =
        sql
            .createQuery(Menu::class) {
                specification?.let(::where)
                orderBy(DEFAULT_SORT)
                select(table.fetch(MenuView::class))
            }.execute()

    /** 查询根节点菜单树 */
    fun listTreeViews(): List<MenuTreeView> =
        sql
            .createQuery(Menu::class) {
                where(table.parentId.isNull())
                orderBy(DEFAULT_SORT)
                select(table.fetch(MenuTreeView::class))
            }.execute()

    /** 判断是否存在子菜单 */
    fun hasChildren(parentId: Long): Boolean =
        sql
            .createQuery(Menu::class) {
                where(table.parentId eq parentId)
                select(table.id)
            }.execute()
            .isNotEmpty()

    /** 按 ID 查询菜单视图 */
    fun loadViewById(id: Long): MenuView? =
        sql
            .createQuery(Menu::class) {
                where(table.id eq id)
                select(table.fetch(MenuView::class))
            }.execute()
            .firstOrNull()

    /** 按角色编码列表查询授权菜单视图 */
    fun listViewsByRoleCodes(roleCodes: List<String>): List<MenuView> {
        if (roleCodes.isEmpty()) {
            return emptyList()
        }
        val menus =
            roleCodes.flatMap { roleCode ->
                sql
                    .createQuery(Role::class) {
                        where(table.code eq roleCode)
                        select(table.fetch(AuthorizationFetchers.ROLE_WITH_MENUS))
                    }.execute()
                    .flatMap { role -> role.menus }
                    .map(::MenuView)
            }
        return menus
            .distinctBy { it.id }
            .sortedWith(compareBy<MenuView> { it.sortingOrder ?: Int.MAX_VALUE }.thenBy { it.id })
    }

    companion object {
        private val DEFAULT_SORT: Sort =
            Sort.by(Sort.Order.asc("sortingOrder"), Sort.Order.asc("id"))
    }
}
